File: src/DXFViewerDX/DXFViewer/DXF/Block.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DXFViewer
{
    public class Block : DXFDotNet.Block
    {
        protected DXFViewer.DXFControl m_Ctrl = null;

        public Block(DXFViewer.DXFControl ctrl) : base(ctrl)
        {
            m_Ctrl = ctrl;
        }

        public override void Add(DXFDotNet.Shape pObj)
        {
            base.Add(pObj);
        }

        public override bool Remove(DXFDotNet.Shape pObj)
        {
            bool bResult = base.Remove(pObj);

            return bResult;
        }

        public override void RemoveAll()
        {
            base.RemoveAll();
        }
    }
}

File: src/DXFViewerDX/DXF2DotNet/Block.cpp
#include "StdAfx.h"
#include "Block.h"
#include "Shape.h"
#include "LineType.h"

using namespace System::Collections;

BEGIN_NS(DXFDotNet)

Block::Block(DXFControl^ ctrl)
{
	m_isHidden = false;
	m_isLock   = false;
	m_vOrigin  = gcnew UnE::Geometry::Vertex2D(0.0, 0.0);

	m_listObj  = gcnew ArrayList();
	m_lineType = gcnew LineType(ctrl);
	m_color = System::Drawing::Color::White;
	m_ctrl = ctrl;

	m_strBlockName = L"";
}

Block::~Block(void)
{
	RemoveAll();
}

void Block::Add(Shape^ pObj)
{
	if (pObj != nullptr)
		m_listObj->Add(pObj);
}

bool Block::Remove(Shape^ pObj)
{
	if (pObj == nullptr) return false;
	pObj->SetBlock(nullptr);

	if (m_listObj->Contains(pObj))
	{
		m_listObj->Remove(pObj);
		return true;
	}

	return false;
}

void Block::RemoveAll()
{
	m_listObj->Clear();
}

void Block::SetLineType(LineType^ lineType)
{
	m_lineType = lineType;
}

LineType^ Block::GetLineType()
{
	return m_lineType;
}

END_NS
